earningsbot.py

```diff
@@ -148,38 +146,57 @@ def send_message(self):
         return webhook.execute()
 
 
-while True:
-    # Don't run on weekends.
-    if datetime.today().weekday() > 4:
+def working_for_the_weekend(today: datetime = datetime.today()) -> bool:
+    """Don't run on weekends."""
+    if today.weekday() > 4:
         logging.info("Skipping run due to weekend.")
-        sleep(3600)
-        continue
+        return True
+
+    return False
 
-    # Get a list of messages on StockTwits.
+
+def generate_messages(poll_interval: int = 300):
+    """Returns a generator that yields messages from StockTwits."""
     URL = "https://api.stocktwits.com/api/2/streams/user/epsguid.json"
     headers = {
         "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:122.0) Gecko/20100101 Firefox/122.0"
     }
     params = {"filter": "all", "limit": 21}
-    logging.info("Getting latest messages...")
-    resp = requests.get(URL, params=params, headers=headers)
 
-    # Skip reporting if this is the first time we've run since a restart.
-    if last_message_id == 0 and FORCED_MESSAGES == 0:
-        logging.info("Not reporting on first run.")
-        last_message_id = resp.json()["messages"][0]["id"]
+    last_message_id = 0
 
-    # Loop over the messages and report on each that hasn't been seen previously.
-    for message in reversed(resp.json()["messages"]):
-        if message["id"] > last_message_id:
-            earnings_publisher = EarningsPublisher(message)
-            if earnings_publisher.earnings:
-                earnings_publisher.send_message()
+    while True:
+        logging.info("Getting latest messages...")
+        resp = requests.get(URL, params=params, headers=headers)
 
-            logging.info(earnings_publisher.title)
+        # Skip reporting if this is the first time we've run since a restart.
+        if last_message_id == 0 and FORCED_MESSAGES == 0:
+            logging.info("Not reporting on first run.")
+            last_message_id = resp.json()["messages"][0]["id"]
+
+        # Loop over the messages and report on each that hasn't been seen previously.
+        for message in reversed(resp.json()["messages"]):
+            if message["id"] > last_message_id:
+                yield message
 
             # Store this message for next time.
             last_message_id = message["id"]
 
-    logging.info("Waiting 5 minutes before next run...")
-    sleep(300)
+        # Sleep for 5 minutes before checking for new messages.
+        logging.info(earnings_publisher.title)
+        logging.info("Waiting 5 minutes before next run...")
+        sleep(poll_interval)
+
+
+if __name__ == "__main__":
+    while True:
+        if working_for_the_weekend():
+            # Sleep for an hour and try again.
+            sleep(3600)
+            continue
+
+        for message in generate_messages():
+            earnings_publisher = EarningsPublisher(message)
+
+            if earnings_publisher.earnings:
+                earnings_publisher.send_message()
```

Comment:
This is a bug ;( after the first check of the weekend, it will never check again.

Reply:
Now with more better fixes.